File: examples/71-action-api/main.go
// examples/71-action-api demonstrates hyperlinks and click actions on shapes.
//
// Shows URL links, slide navigation (first/last/next/prev/specific slide),
// email links, hover actions, and how to attach hyperlinks to text runs.
//
// Run with: go run ./examples/71-action-api/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	log "github.com/djinn-soul/gopptx/pkg/stdlog"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/action"
	"github.com/djinn-soul/gopptx/pkg/pptx/elements"
)

const (
	outputDir  = "examples/output"
	outputFile = "71_action_api.pptx"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output directory: %w", err)
	}

	slides := []pptx.SlideContent{
		buildURLHyperlinkSlide(),
		buildNavigationSlide(),
		buildJumpToSlideSlide(),
		buildEmailHyperlinkSlide(),
		buildHoverActionSlide(),
		buildTextRunHyperlinkSlide(),
	}

	outputPath := filepath.Join(outputDir, outputFile)
	data, err := pptx.CreateWithSlides("Action API Demo", slides)
	if err != nil {
		return fmt.Errorf("create presentation: %w", err)
	}
	if err := os.WriteFile(outputPath, data, 0o600); err != nil {
		return fmt.Errorf("write file: %w", err)
	}

	log.Printf("Generated %s\n", outputPath)
	return nil
}

func buildURLHyperlinkSlide() pptx.SlideContent {
	urlLink := action.NewHyperlink(
		action.HyperlinkURL("https://github.com/djinn-soul/gopptx"),
	).
		WithTooltip("Visit gopptx on GitHub").
		WithHighlightClick(true)

	urlShape := pptx.NewRectangle(1, 1.5, 5, 1).
		WithFill(pptx.NewShapeFill("4472C4")).
		WithText("Click to open GitHub (URL hyperlink)").
		WithClickAction(urlLink)

	return pptx.NewSlide("URL Hyperlink on Shape").
		AddShape(urlShape).
		AddBullet("Shapes can carry click actions via WithClickAction.")
}

func buildNavigationSlide() pptx.SlideContent {
	nextBtn := pptx.NewRectangle(0.5, 4.5, 2, 0.8).
		WithFill(pptx.NewShapeFill("9BBB59")).
		WithText("Next Slide").
		WithClickAction(action.NewHyperlink(action.HyperlinkNextSlide()))

	prevBtn := pptx.NewRectangle(3, 4.5, 2, 0.8).
		WithFill(pptx.NewShapeFill("C0504D")).
		WithText("Prev Slide").
		WithClickAction(action.NewHyperlink(action.HyperlinkPreviousSlide()))

	firstBtn := pptx.NewRectangle(5.5, 4.5, 2, 0.8).
		WithFill(pptx.NewShapeFill("F79646")).
		WithText("First Slide").
		WithClickAction(action.NewHyperlink(action.HyperlinkFirstSlide()))

	lastBtn := pptx.NewRectangle(0.5, 5.5, 2, 0.8).
		WithFill(pptx.NewShapeFill("8064A2")).
		WithText("Last Slide").
		WithClickAction(action.NewHyperlink(action.HyperlinkLastSlide()))

	endShowBtn := pptx.NewRectangle(3, 5.5, 2, 0.8).
		WithFill(pptx.NewShapeFill("333333")).
		WithText("End Show").
		WithClickAction(action.NewHyperlink(action.HyperlinkEndShow()))

	return pptx.NewSlide("Slide Navigation Actions").
		AddShape(nextBtn).
		AddShape(prevBtn).
		AddShape(firstBtn).
		AddShape(lastBtn).
		AddShape(endShowBtn).
		AddBullet("Use HyperlinkNextSlide, PreviousSlide, FirstSlide, LastSlide, EndShow.")
}

func buildJumpToSlideSlide() pptx.SlideContent {
	jumpBtn := pptx.NewRectangle(1, 2, 4, 1).
		WithFill(pptx.NewShapeFill("4BACC6")).
		WithText("Jump to Slide 5").
		WithClickAction(action.NewHyperlink(action.HyperlinkSlide(5)))

	return pptx.NewSlide("Jump to Specific Slide").
		AddShape(jumpBtn).
		AddBullet("HyperlinkSlide(n) jumps to slide number n in the show.")
}

func buildEmailHyperlinkSlide() pptx.SlideContent {
	emailLink := action.NewHyperlink(
		action.HyperlinkEmailWithSubject("hello@example.com", "Question about gopptx"),
	).WithTooltip("Send us an email")

	emailShape := pptx.NewRectangle(1, 2, 6, 1).
		WithFill(pptx.NewShapeFill("EBF1DE")).
		WithText("Click to email hello@example.com").
		WithClickAction(emailLink)

	return pptx.NewSlide("Email Hyperlink").
		AddShape(emailShape).
		AddBullet("HyperlinkEmailWithSubject sets the mailto: href with a subject.")
}

func buildHoverActionSlide() pptx.SlideContent {
	hoverShape := pptx.NewRectangle(1, 2, 6, 1).
		WithFill(pptx.NewShapeFill("FDE9D9")).
		WithText("Hover over me!").
		WithHoverAction(action.NewHyperlink(action.HyperlinkURL("https://example.com")))

	return pptx.NewSlide("Hover Action").
		AddShape(hoverShape).
		AddBullet("Shapes can also have hover actions via WithHoverAction.")
}

func buildTextRunHyperlinkSlide() pptx.SlideContent {
	linkRun := elements.NewRun("click here").WithHyperlink(
		action.NewHyperlink(action.HyperlinkURL("https://pkg.go.dev")),
	)
	plainRun1 := elements.NewRun("Visit Go docs – ")
	plainRun2 := elements.NewRun(" – for package documentation.")

	return pptx.NewSlide("Hyperlink on Text Run").
		AddBulletRuns([]elements.Run{plainRun1, linkRun, plainRun2}).
		AddBullet("Text runs support WithHyperlink for inline links.")
}

File: examples/71-action-api/main.py
"""Document hyperlinks and click-action patterns on shapes.

This example demonstrates:
- Adding shapes with add_shape() and labels explaining action concepts
- Documenting URL hyperlinks, slide navigation, email links, and hover actions
- Using add_bullet_slide() and add_textbox() to explain each action type
- Full click/hover action wiring is available in the Go API
- Python slides document concepts and show shape placement for those actions
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation
from gopptx.constants import ShapeType
from gopptx.presentation.slides import SlideLayoutType
from gopptx.schemas import Inches


def _add_url_hyperlink_slide(prs: Presentation) -> None:
    """Slide describing URL hyperlink actions."""
    prs.add_bullet_slide(
        "URL Hyperlink on Shape",
        [
            "Shapes can carry click actions (click-to-URL).",
            "URL: https://github.com/djinn-soul/gopptx",
            "In Go: shape.WithClickAction(action.NewHyperlink(...))",
            "Tooltip and highlight-on-click are configurable.",
        ],
    )


def _add_navigation_slide(prs: Presentation) -> None:
    """Slide describing slide navigation actions."""
    idx = prs.add_slide(
        "Slide Navigation Actions", layout=SlideLayoutType.TITLE_ONLY
    ).index
    labels = [
        ("Next Slide", "4CAF50", Inches(0.5), Inches(1.5)),
        ("Prev Slide", "F44336", Inches(3.0), Inches(1.5)),
        ("First Slide", "FF9800", Inches(5.5), Inches(1.5)),
        ("Last Slide", "9C27B0", Inches(0.5), Inches(3.0)),
        ("End Show", "455A64", Inches(3.0), Inches(3.0)),
    ]
    for label, color, x, y in labels:
        prs.add_shape(
            idx,
            ShapeType.RECTANGLE,
            (x, y, Inches(2.2), Inches(0.8)),
            text=label,
            properties={"fill_color": color},
        )
    prs.add_textbox(
        idx,
        Inches(0.5),
        Inches(4.2),
        Inches(9),
        Inches(0.6),
        text="In Go: WithClickAction(action.NewHyperlink(HyperlinkNextSlide()))",
    )


def _add_jump_to_slide_slide(prs: Presentation) -> None:
    """Slide describing jump-to-specific-slide action."""
    idx = prs.add_slide(
        "Jump to Specific Slide", layout=SlideLayoutType.TITLE_ONLY
    ).index
    prs.add_shape(
        idx,
        ShapeType.RECTANGLE,
        (Inches(1), Inches(2), Inches(4), Inches(1)),
        text="Jump to Slide 5",
        properties={"fill_color": "4BACC6"},
    )
    prs.add_textbox(
        idx,
        Inches(0.5),
        Inches(3.5),
        Inches(9),
        Inches(0.6),
        text="In Go: WithClickAction(action.NewHyperlink(action.HyperlinkSlide(5)))",
    )


def _add_email_hyperlink_slide(prs: Presentation) -> None:
    """Slide describing email hyperlink action."""
    idx = prs.add_slide("Email Hyperlink", layout=SlideLayoutType.TITLE_ONLY).index
    prs.add_shape(
        idx,
        ShapeType.ROUNDED_RECTANGLE,
        (Inches(1), Inches(2), Inches(6), Inches(1)),
        text="Click to email hello@example.com",
        properties={"fill_color": "EBF1DE"},
    )
    prs.add_textbox(
        idx,
        Inches(0.5),
        Inches(3.5),
        Inches(9),
        Inches(0.8),
        text="In Go: HyperlinkEmailWithSubject('hello@example.com', 'Subject')",
    )


def _add_hover_action_slide(prs: Presentation) -> None:
    """Slide describing hover action."""
    idx = prs.add_slide("Hover Action", layout=SlideLayoutType.TITLE_ONLY).index
    prs.add_shape(
        idx,
        ShapeType.RECTANGLE,
        (Inches(1), Inches(2), Inches(6), Inches(1)),
        text="Hover over me!",
        properties={"fill_color": "FDE9D9"},
    )
    prs.add_textbox(
        idx,
        Inches(0.5),
        Inches(3.5),
        Inches(9),
        Inches(0.6),
        text="In Go: shape.WithHoverAction(action.NewHyperlink(...))",
    )


def main() -> None:
    """Create presentation documenting hyperlink and action concepts."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    output_path = output_dir / "71_action_api.pptx"

    with Presentation.new("Action API Demo") as prs:
        _add_url_hyperlink_slide(prs)
        _add_navigation_slide(prs)
        _add_jump_to_slide_slide(prs)
        _add_email_hyperlink_slide(prs)
        _add_hover_action_slide(prs)

        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print("Demonstrated: URL hyperlinks, slide navigation buttons, jump-to-slide,")
    print("  email hyperlinks, and hover actions via shape placement + labels")


if __name__ == "__main__":
    main()
